br/pkg/backup/prepare_snap/stream.go

```diff
@@ -0,0 +1,212 @@
+	}
+	msg, err := p.cli.Recv()
+	if err != nil {
+		return errors.Annotate(err, "failed to recv the initialize lease result")
+	}
+	if msg.Ty != brpb.PrepareSnapshotBackupEventType_UpdateLeaseResult {
+		return errors.Errorf("unexpected type of response during creating lease loop: it is %s", msg.Ty)
+	}
+	p.serverStream = utils.AsyncStreamBy(p.cli.Recv)
+	p.clientLoopHandle.Go(func() error { return p.clientLoop(ctx, dur) })
+	return nil
+}
+
+func (p *prepareStream) onResponse(ctx context.Context, res utils.Result[*brpb.PrepareSnapshotBackupResponse]) error {
+	if err := res.Err; err != nil {
+		return err
+	}
+	resp := res.Item
+	logutil.CL(ctx).Debug("received response", zap.Stringer("resp", resp))
+	evt, needDeliver := p.convertToEvent(resp)
+	if needDeliver {
+		logutil.CL(ctx).Debug("generating internal event", zap.Stringer("event", evt))
+		p.output <- evt
+	}
+	return nil
+}
+
+func (p *prepareStream) stopClientLoop(ctx context.Context) error {
+	p.stopBgTasks()
+	err := p.cli.Send(&brpb.PrepareSnapshotBackupRequest{
+		Ty: brpb.PrepareSnapshotBackupRequestType_Finish,
+	})
+	if err != nil {
+		return errors.Annotate(err, "failed to send finish request")
+	}
+recv_loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case res, ok := <-p.serverStream:
+			err := p.onResponse(ctx, res)
+			if err == io.EOF || !ok {
+				logutil.CL(ctx).Info("close loop done.", zap.Uint64("store", p.storeID), zap.Bool("is-chan-closed", !ok))
+				break recv_loop
+			}
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return p.clientLoopHandle.Wait()
+}
+
+func (p *prepareStream) clientLoop(ctx context.Context, dur time.Duration) error {
+	ticker := time.NewTicker(dur / 4)
+	lastSuccess := time.Unix(0, 0)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			logutil.CL(ctx).Info("client loop exits.", zap.Uint64("store", p.storeID))
+			return nil
+		case res := <-p.serverStream:
+			if err := p.onResponse(ctx, res); err != nil {
+				p.sendErr(errors.Annotate(err, "failed to recv from the stream"))
+				return err
+			}
+		case <-ticker.C:
+			err := p.cli.Send(&brpb.PrepareSnapshotBackupRequest{
+				Ty:             brpb.PrepareSnapshotBackupRequestType_UpdateLease,
+				LeaseInSeconds: uint64(dur.Seconds()),
+			})
+			if err != nil {
+				log.Warn("failed to update the lease loop", logutil.ShortError(err))
+				if time.Since(lastSuccess) > dur {
+					err := errors.Annotate(err, "too many times failed to update the lease, it is probably expired")
+					p.output <- event{
+						ty:      eventMiscErr,
+						storeID: p.storeID,
+						err:     err,
+					}
+					return err
+				}
+			} else {
+				lastSuccess = time.Now()
+			}
+		}
+	}
+}
+
+func (p *prepareStream) sendErr(err error) {
+	p.output <- event{
+		ty:      eventMiscErr,
+		storeID: p.storeID,
+		err:     err,
+	}
+}
+
+func (p *prepareStream) convertToEvent(resp *brpb.PrepareSnapshotBackupResponse) (event, bool) {
+	switch resp.Ty {
```

Comment:
What's the response type of `PrepareSnapshotBackupRequestType_Finish`?

Reply:
Update lease result.